Updated AQI calculation for HM3301 to the new standard

optimusprimespace opened the pull request:
# What does this implement/fix?

<!-- Quick description and explanation of changes -->
Updated the AQI calculation formula to the newest 2024 issued document standard that EPA published [Technical Assistance Document for the Reporting of Daily Air Quality – the Air Quality Index (AQI) , 2024] https://document.airnow.gov/technical-assistance-document-for-the-reporting-of-daily-air-quailty.pdf
It's only a small fix, but a breaking code change, because it now measures the AQI index differently that it used to.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [X] ESP32
- [x] ESP32 IDF
- [x] ESP8266
- [x] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

```

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

optimusprimespace committed:
Updated AQI to the new standard

optimusprimespace committed:
clang-format update

optimusprimespace committed:
clang-fix

optimusprimespace committed:
clang-fix (small)

optimusprimespace committed:
Merge branch 'esphome:dev' into hm3301

optimusprimespace committed:
Merge branch 'esphome:dev' into hm3301

optimusprimespace committed:
Merge branch 'esphome:dev' into hm3301

optimusprimespace committed:
Merge branch 'esphome:dev' into hm3301

bdraco committed:
Merge branch 'dev' into hm3301

bdraco commented on esphome/components/hm3301/aqi_calculator.h:
```diff
@@ -16,16 +16,15 @@ class AQICalculator : public AbstractAQICalculator {
-  int pm2_5_calculation_grid_[AMOUNT_OF_LEVELS][2] = {{0, 12},    {13, 35},   {36, 55},  {56, 150},
-                                                      {151, 250}, {251, 350}, {351, 500}};
+  int pm2_5_calculation_grid_[AMOUNT_OF_LEVELS][2] = {{0, 9},    {10, 35},   {36, 55},
+                                                      {56, 125}, {126, 225}, {256, INT_MAX}};
```
```suggestion
                                                      {56, 125}, {126, 225}, {226, INT_MAX}};
```

This looks like there is a typo

optimusprimespace replied:
Thanks for finding the typo!

optimusprimespace committed:
Fixed Typo

Acoording to the doc (https://document.airnow.gov/technical-assistance-document-for-the-reporting-of-daily-air-quailty.pdf) on page 19 it should be 226 not 256

optimusprimespace committed:
Merge branch 'esphome:dev' into hm3301

optimusprimespace commented:
@bdraco My ultimate goal is to create a new component (separate from this) for AQI where you give the component values from the sensors you want (only if they have the necessary measured values like CO or PM10). Then the component will process the data according to the documentation and give you an easy to read air quality index.
